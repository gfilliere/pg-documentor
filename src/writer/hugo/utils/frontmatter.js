// @flow
import yaml from 'js-yaml';
import toml from 'toml-js';

export const FRONTMATTER_YAML = 'YAML';
export const FRONTMATTER_TOML = 'TOML';
export const FRONTMATTER_JSON = 'JSON';

type frontMatterFormat = 'YAML' | 'TOML' | 'JSON';

/**
 * Creates a hugo compatible frontmatter
 * https://gohugo.io/content-management/front-matter/
 */
export default function createFrontmatter(
  data: any,
  format: frontMatterFormat = FRONTMATTER_TOML,
) {
  switch (format) {
    case FRONTMATTER_YAML:
      return `---
${yaml.safeDump(data).trim()}
---
`;
    case FRONTMATTER_TOML:
      return `+++
${toml.dump(data).trim()}
+++
`;
    case FRONTMATTER_JSON:
      return `${JSON.stringify(data, null, 2).trim()}`;
    default:
      throw new Error(`Unsupported format: ${format}`);
  }
}
