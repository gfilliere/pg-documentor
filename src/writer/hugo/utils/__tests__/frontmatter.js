import frontmatter, {
  FRONTMATTER_JSON,
  FRONTMATTER_YAML,
  FRONTMATTER_TOML,
} from '../frontmatter';

const sampleFrontMatter = {
  title: 'spf13-vim 3.0 release and new website',
  description:
    'spf13-vim is a cross platform distribution of vim plugins and resources for Vim.',
  tags: ['.vimrc', 'plugins', 'spf13-vim', 'vim'],
  date: '2012-04-06',
  categories: ['Development', 'VIM'],
  slug: 'spf13-vim-3-0-release-and-new-website',
};

test('generates valid toml', () => {
  expect(frontmatter(sampleFrontMatter, FRONTMATTER_TOML)).toMatchSnapshot();
});

test('generates valid yaml', () => {
  expect(frontmatter(sampleFrontMatter, FRONTMATTER_YAML)).toMatchSnapshot();
});

test('generates valid JSON', () => {
  expect(frontmatter(sampleFrontMatter, FRONTMATTER_JSON)).toMatchSnapshot();
});
