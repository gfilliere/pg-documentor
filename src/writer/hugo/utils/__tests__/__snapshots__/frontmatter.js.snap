// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates valid JSON 1`] = `
"{
  \\"title\\": \\"spf13-vim 3.0 release and new website\\",
  \\"description\\": \\"spf13-vim is a cross platform distribution of vim plugins and resources for Vim.\\",
  \\"tags\\": [
    \\".vimrc\\",
    \\"plugins\\",
    \\"spf13-vim\\",
    \\"vim\\"
  ],
  \\"date\\": \\"2012-04-06\\",
  \\"categories\\": [
    \\"Development\\",
    \\"VIM\\"
  ],
  \\"slug\\": \\"spf13-vim-3-0-release-and-new-website\\"
}"
`;

exports[`generates valid toml 1`] = `
"+++
title = \\"spf13-vim 3.0 release and new website\\"
description = \\"spf13-vim is a cross platform distribution of vim plugins and resources for Vim.\\"
tags = [\\".vimrc\\", \\"plugins\\", \\"spf13-vim\\", \\"vim\\"]
date = \\"2012-04-06\\"
categories = [\\"Development\\", \\"VIM\\"]
slug = \\"spf13-vim-3-0-release-and-new-website\\"
+++
"
`;

exports[`generates valid yaml 1`] = `
"---
title: spf13-vim 3.0 release and new website
description: >-
  spf13-vim is a cross platform distribution of vim plugins and resources for
  Vim.
tags:
  - .vimrc
  - plugins
  - spf13-vim
  - vim
date: '2012-04-06'
categories:
  - Development
  - VIM
slug: spf13-vim-3-0-release-and-new-website
---
"
`;
